src/utils/auth/auth.provider.tsx

```diff
@@ -0,0 +1,481 @@
+
+        completedAuthRequests.forEach((completedAuthRequest, i) => {
+          const completedAuthRequestIndex = nextAuthRequests.findIndex(
+            (authRequest) => authRequest.authID === completedAuthRequest.authID
+          );
+
+          if (completedAuthRequestIndex === -1) {
+            console.warn(
+              `Could not find AuthRequest with authID="${completedAuthRequestsStatus}"`
+            );
+
+            return;
+          }
+
+          nextAuthRequests[completedAuthRequestIndex] = {
+            ...nextAuthRequests[completedAuthRequestIndex],
+            completedAt: Date.now(),
+            status: completedAuthRequestsStatus[i]
+          };
+
+          if (i === 0)
+            nextLastCompletedAuthRequest =
+              nextAuthRequests[completedAuthRequestIndex];
+        });
+
+        // Find the index of the next "pending" `AuthRequest`, or keep it unchanged if there are none left:
+        let nextCurrentAuthRequestIndex = currentAuthRequestIndex;
+
+        do {
+          nextCurrentAuthRequestIndex =
+            (nextCurrentAuthRequestIndex + 1) % nextAuthRequests.length;
+        } while (
+          nextCurrentAuthRequestIndex !== currentAuthRequestIndex &&
+          nextAuthRequests[nextCurrentAuthRequestIndex].status !== "pending"
+        );
+
+        if (nextCurrentAuthRequestIndex === currentAuthRequestIndex) {
+          nextCurrentAuthRequestIndex = nextAuthRequests.length;
+        }
+
+        return {
+          authRequests: nextAuthRequests,
+          currentAuthRequestIndex: nextCurrentAuthRequestIndex,
+          lastCompletedAuthRequest: nextLastCompletedAuthRequest
+        };
+      });
+    },
+    [authRequests]
+  );
+
+  useEffect(() => {
+    log(LOG_GROUP.AUTH, "Auth popup initialized. Waiting for AuthRequests");
+
+    isomorphicOnMessage("auth_request", (authRequestMessage) => {
+      log(LOG_GROUP.AUTH, "auth_request =", authRequestMessage);
+
+      // UnlockAuthRequests are not enqueued as those are simply used to open the popup to prompt users to enter their
+      // password and wait for the wallet to unlock:
+
+      if (!authRequestMessage?.data) {
+        console.warn("auth_request without data");
+        return;
+      }
+
+      const authRequest = authRequestMessage.data;
+
+      setAuthRequestContextState((prevAuthRequestContextState) => {
+        const {
+          authRequests,
+          currentAuthRequestIndex,
+          lastCompletedAuthRequest
+        } = prevAuthRequestContextState;
+
+        // TODO: Additional considerations when enqueueing new `AuthRequest`s:
+        //
+        // - `AuthRequest`s could be grouped by domain and/or tab. This means the auth popup could/should provide a
+        //   domain/tab selector and automatically select the active tab when the user switches it.
+        //
+        // - Separate `signDataItem`requests could/should automatically be combined into a single `batchSignDataItem`
```

Comment:
I love all suggestions
On the many signDataItem -> batch I agree to showing them as a batch as long as we return each resolution individually to honor the API calls